### PR TITLE
drivers: modem: hl7800: use new work queue API

### PR DESCRIPTION
**Pull request by rerickson1**
Implement new work queue API changes.

addresses: https://github.com/zephyrproject-rtos/zephyr/issues/33925

### CODE REVIEW

File: drivers/modem/hl7800.c [hunk 1]
```diff
@@ -2324,10 +2330,14 @@ static bool on_cmd_network_report_query(struct net_buf **buf, uint16_t len)
 		set_network_state(strtol(val, NULL, 0));
 
 		/* start work to adjust iface */
-		k_delayed_work_cancel(&ictx.iface_status_work);
-		k_delayed_work_submit_to_queue(&hl7800_workq,
-					       &ictx.iface_status_work,
-					       IFACE_WORK_DELAY);
+		rc = k_work_cancel_delayable(&ictx.iface_status_work);
```
Comment (pabigot) [inline drivers/modem/hl7800.c:2333]: If you're going to unconditionally reschedule, there's no point in trying to cancel first.  Reschedule will replace any pending work.

File: drivers/modem/hl7800.c [hunk 2]
```diff
@@ -2466,9 +2477,13 @@ static bool on_cmd_network_report(struct net_buf **buf, uint16_t len)
 	/* keep HL7800 awake because we want to process the network state soon */
 	allow_sleep(false);
 	/* start work to adjust iface */
-	k_delayed_work_cancel(&ictx.iface_status_work);
-	k_delayed_work_submit_to_queue(&hl7800_workq, &ictx.iface_status_work,
-				       IFACE_WORK_DELAY);
+	rc = k_work_cancel_delayable(&ictx.iface_status_work);
```
Comment (pabigot) [inline drivers/modem/hl7800.c:2480]: Same here.

File: drivers/modem/hl7800.c [hunk 3]
```diff
@@ -2781,9 +2796,9 @@ static void sock_read(struct net_buf **buf, uint16_t len)
 
 	if (sock->error) {
 		/* cancel notif work and restart */
-		k_delayed_work_cancel(&sock->notif_work);
-		k_delayed_work_submit_to_queue(&hl7800_workq, &sock->notif_work,
-					       MDM_SOCK_NOTIF_DELAY);
+		k_work_cancel_delayable(&sock->notif_work);
```
Comment (pabigot) [inline drivers/modem/hl7800.c:2799]: same here.

**Review by pabigot — CHANGES_REQUESTED** [review]
Any place where you cancel then immediately reschedule you don't need the cancel.  That should be changed, even though it isn't new.

I'm not sure it's worth emitting a diagnostic on failure to cancel if the code isn't going to do anything about the failure, but that's maintainer choice.

**Review by pabigot — CHANGES_REQUESTED** [review]
Found another problem: all the work handler functions that construct a pointer to a container from the passed pointer to a work item need to be updated: you can't assume the work structure is at the start of the `struct k_work_delayable`, and you're not allowed to inspect into that structure.

See https://github.com/zephyrproject-rtos/zephyr/blob/a865b1bb49f1b575d876f102a8e0ff4dbeb7cdf7/tests/kernel/workq/work/src/main.c#L967 for an example of how to do it properly.  The advice in #33104 has been updated to explicitly mention this.

**Comment by rerickson1** [comment]
> Found another problem: all the work handler functions that construct a pointer to a container from the passed pointer to a work item need to be updated: you can't assume the work structure is at the start of the `struct k_work_delayable`, and you're not allowed to inspect into that structure.
> 
> See
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/a865b1bb49f1b575d876f102a8e0ff4dbeb7cdf7/tests/kernel/workq/work/src/main.c#L967
> 
> for an example of how to do it properly. The advice in #33104 has been updated to explicitly mention this.

Only need rework for the work handlers for delayed work correct?

File: drivers/modem/hl7800.c [hunk 4]
```diff
@@ -2618,8 +2621,10 @@ static bool on_cmd_sock_error_code(struct net_buf **buf, uint16_t len)
 static void sock_notif_cb_work(struct k_work *work)
 {
 	struct hl7800_socket *sock = NULL;
+	struct k_work_delayable *dwork;
 
-	sock = CONTAINER_OF(work, struct hl7800_socket, notif_work);
+	dwork = k_work_delayable_from_work(work);
+	sock = CONTAINER_OF(dwork, struct hl7800_socket, notif_work);
 
 	if (!sock) {
```
Comment (pabigot) [inline drivers/modem/hl7800.c:2629]: In what situation would this test succeed?   `CONTAINER_OF` returns an appropriately cast pointer that is offset by some number of bytes from the `work` pointer.  It's not going to be null.

**Review by pabigot — APPROVED** [review]
This seems fine for transformation of the existing code.  I have one comment about a test that seems wrong, but it's not a new problem.

AFAICT this module is not SMP safe; below are some comments related to fixing that.  These also aren't new, but are worth mentioning.

These three work items (handlers) can be cancelled, with these potential problems in SMP systems or preemptive threads where the cancellation can fail to succeed:

* `allow_sleep_work` (`allow_sleep_work_callback`)
* `rssi_query_work` (`hl7800_rssi_query_work`)
* `notif_work` (`sock_notif_cb_work`)

For `allow_sleep_work`: if the work handler is in state `RUNNING` when `allow_sleep(false)` is invoked the cancel can fail, and the modem will wake contrary to the caller's expectations.

`rssi_query_work` may have a similar problem; I haven't looked closely.

`notif_work` likely can do the same thing, but since its work function takes a lock the problem (if it exists) might be eliminated by taking `hl7800_lock()` before attempting the cancel, and if the cancel fails setting some flag balue in `*sock` (e.g. `sock->type == SOCK_NOT_A_VALID_TYPE`) and checking for that within the locked section of the work handler (fast-exiting with no action if it's discovered).

**Review by jukkar — APPROVED** [review]

File: drivers/modem/hl7800.c (hunk 4 above)
Comment (rerickson1) [reply drivers/modem/hl7800.c]: Agree, this check isn't needed.

**Pull request merged**